### PR TITLE
scripts: don't build archives when pages haven't been changed

### LINKED ISSUE #11823: Archive checksums change on every commit, even if the files inside are not changed
The checksum file was added in #9669 to enable clients to download archives only if they have been changed. However, that is not possible as of now, because zip archives store the modification time of each file. Since Git does not record these timestamps, `mtime` is set to the current date and time on checkout. Different metadata means different hashes, and so the client downloads the exact same archive once again.

Knowing the checksums, clients can only know if there has been a commit since the last cache update. Even if the commit in question changes the README, for example.

[Here is an example.](https://github.com/tldr-pages/tldr-pages.github.io/commit/e8e5168dcf920a6bbe1776defcdd5baa75085cd5#diff-58330d3e46dadb5b59c35c4c077adfc213c9c34f3788c7e5fbe6035ef11d9f5e) The only change in that commit was the deletion of `id_ed25519_tldr_asset_upload.enc`, yet all the hashes have changed.

---

I tried to create a zip without modification times, but I don't think that's possible.
- `zip` has no option to change that.
- `7z` does have it (`7z a -mtm=off`), but using it on a zip archive produces an error:
```
System ERROR:
E_INVALIDARG
```

---

The best way to fix this would be to not build archives when the files have not been changed ~, but I have no idea where to even begin.~ I tried to do that (#11825)

### PR DESCRIPTION
**Pull request by acuteenvy**
`git --name-only @^ @` outputs paths of all changed files in the latest commit.
`cut -d'/' -f1` outputs the first element of the path (i.e. `pages.lang`).
Before building an archive, `build.sh` now checks if the directory is in that list.

---

I tested this in a GitHub repo, and it correctly builds only the changed archives.

Closes #11823.

### CODE REVIEW

**Review by kbdharun — COMMENTED** [review]
Thanks for your contribution, I tested the changes of this PR [here](https://github.com/kbdkorg/tldr/actions?query=branch%3Aacuteenvy%2Fconsistent-sums).  

While the changes worked as intended, I noticed an issue since we don't build new PDFs or assets when pages aren't modified the deploy step would fail stating that the branch is already in sync with the main which fails the whole action, thus notifying us (maintainer who merged the PR) on every instance a non-page change is committed to `tldr:main`. 

While this is fine, I would suggest updating it to a quieter method where the action wouldn't downright fail when there aren't any changes in pages.

**Edit.** I tested out a method for it, will apply the same here.

**Review by kbdharun — APPROVED** [review]
Done, added a conditional check so that when no pages are modified, the action won't fail instead it will tell "No changes to deploy." in the actions log. Tested it too. I think this PR is GTG from my side, feel free to review my changes @acuteenvy, @sebastiaanspeck.

**Comment by acuteenvy** [comment]
I moved the check to the top, since there's no need to run `initalize` when there are no changes to commit. That also removes the need to check if the full archive exists - it will always exist when something was modified in `pages*`.

File: scripts/deploy.sh [hunk 1]
```diff
@@ -30,21 +30,25 @@ function initialize {
...
-  echo "Assets (pages archive, index) deployed to static site."
+  # Check if there are changes before committing and pushing.
+  if git diff --quiet; then
+    echo "No changes to deploy."
+  else
+    git add -A
+    git commit -m "[GitHub Actions] uploaded assets after commit tldr-pages/tldr@${GITHUB_SHA}"
+    git push -q
+    echo "Assets (pages archive, index) deployed to the static site."
+  fi
```
Comment (acuteenvy) [inline scripts/deploy.sh:51]:
```suggestion
  # Check if there are changes before committing and pushing.
  if git diff --quiet; then
    echo "No changes to deploy."
    return
  fi

  git add -A
  git commit -m "[GitHub Actions] uploaded assets after commit tldr-pages/tldr@${GITHUB_SHA}"
  git push -q
  echo "Assets (pages archive, index) deployed to the static site."
```

Comment (kbdharun) [reply scripts/deploy.sh:51]: Yeah, this looks much better. Minor nitpick in the commit title along with archive and index can you add checksum too?

Comment (acuteenvy) [reply scripts/deploy.sh:51]: > Yeah, this looks much better. Minor nitpick in the commit title along with archive and index can you add checksum too?

What do you mean? Add "don't compute checkums" to the commit message? You can change the commit message before merging if you want to.

**Comment by acuteenvy** [comment]
Since the full archive will always be built when a page is changed, putting this in the `initialize` function should also work. I haven't tested it though.
```bash
if [[ -f $TLDR_ARCHIVE ]]; then
  echo "No changes to deploy."
  exit 0
fi
```

Then the `-f` checks for the index and archive in `upload_assets` can be removed.

File: scripts/deploy.sh (hunk 1 above)
Comment (kbdharun) [reply scripts/deploy.sh:51]: > > Yeah, this looks much better. Minor nitpick in the commit title along with archive and index can you add checksum too?
> 
> What do you mean? Add "don't compute checkums" to the commit message? You can change the commit message before merging if you want to.

Sorry, I confused the commit message and log, I meant updating the log's message to `echo "Assets (pages archive, index and checksum) deployed to the static site.`

**Comment by kbdharun** [comment]
> Since the full archive will always be built when a page is changed, putting this in the initialize function should also work. No need to even clone the git repo. I haven't tested it though.

Will test it out and see how it works.

**Review by kbdharun — APPROVED** [review]

**Pull request merged**